ai/src/llm_infer.py

```diff
@@ -1,32 +1,56 @@
 import os
+import concurrent.futures
 
 from openai import OpenAI
 import pandas as pd
-
+from loguru import logger
 
 MODEL_NAME = os.environ["OPENAI_MODEL_NAME"]
 
 client = OpenAI(api_key=os.environ["OPENAI_API_KEY"])
 
 
-def run_prompt(prompt: str, inputs: pd.DataFrame) -> pd.DataFrame:
-    if prompt.count("{{document}}") != 1:
-        raise ValueError("Prompt must contain exactly one {{document}} placeholder")
-
-    outputs = []
+def run_prompt(prompt: str, inputs: pd.DataFrame, max_workers: int = 10) -> pd.DataFrame:
+    """Runs the given prompt for each input document in the DataFrame using multithreading.
 
-    for _, row in inputs.iterrows():
-        prompt_with_document = prompt.replace("{{document}}", row["inputs"])
+    Be aware that the function will raise an error if openai api returns an error on any of the documents.
 
-        response = client.chat.completions.create(
-            model=MODEL_NAME,
-            messages=[{"role": "user", "content": prompt_with_document}],
-        )
+    Args:
+        prompt: The prompt template string. Must contain "{{document}}".
+        inputs: DataFrame with an "inputs" column containing the documents.
 
-        model_output = response.choices[0].message.content
+    Returns:
+        The input DataFrame with an added "outputs" column containing the model responses.
 
-        outputs.append(model_output)
-
-    inputs["outputs"] = outputs
+    Raises:
+        ValueError: If the prompt does not contain exactly one "{{document}}" placeholder.
+    """
+    if prompt.count("{{document}}") != 1:
+        raise ValueError("Prompt must contain exactly one {{document}} placeholder")
 
+    def _get_openai_response(input_document: str) -> str:
+        """Helper function to call OpenAI API for a single document.
+
+        Args:
+            input_document: The input document to process.
+
+        Returns:
+            The model output for the input document.
+        """
+        prompt_with_document = prompt.replace("{{document}}", input_document)
+        try:
+            response = client.chat.completions.create(
+                model=MODEL_NAME,
+                messages=[{"role": "user", "content": prompt_with_document}],
+            )
+            model_output = response.choices[0].message.content
+            return model_output if model_output is not None else ""
+        except Exception as e:
+            logger.error(f"Error processing document: {e}")
+            raise ValueError(f"Error processing document: {e}, input: {input_document}")
```

Comment:
honestly I think that writing functions inside other functions is a bad pattern. It 
- (a) makes the `run_prompt` function harder to understand because it gets 3 times longer and with large nestedness inside. 
- (b) doesn't allow us to easily reuse `_get_openai_response` if we need it in some other place.


I'd suggest having "run_promp_on_df(prompt, df, max_workers)" and "_run_prompt_on_document(prompt, document)"